tools/wavetool.py

```diff
@@ -0,0 +1,151 @@
+    # here we set the digital zero threshold to 100, and cut samples below 100
+    # of two sides, this has no negative impact to processing.
+    DIGITAL_ZERO_THRESHOLD = 100
+    wave_mono = wave[:,0]
+    left_idx = 0
+    right_idx = wave_mono.shape[0] - 1
+    while True:
+        if abs(wave_mono[left_idx]) > DIGITAL_ZERO_THRESHOLD:
+            break
+        left_idx = left_idx + 1
+    while True:
+        if abs(wave_mono[right_idx]) > DIGITAL_ZERO_THRESHOLD:
+            break
+        right_idx = right_idx - 1
+    return wave[left_idx:right_idx,:], left_idx
+
+# float point binary comparison is not supported, and will not be supported
+# check recorded wave through smart amplifier component
+def compare_wav_smart_amp(wave, ref_wave, fs):
+    trimed_ref_wave, _ = trim_wave(ref_wave)
+    # compare the first two channel
+    trimed_wave_ch_0_1, delay1 = trim_wave(wave[:,0:2])
+    trimed_ref_wave = trimed_ref_wave[0:trimed_wave_ch_0_1.shape[0],:]
+    compare_result0_1 = np.array_equal(trimed_ref_wave, trimed_wave_ch_0_1)
+    # compare the second two channel
+    trimed_wave_ch_2_3, delay2 = trim_wave(wave[:,2:4])
+    trimed_ref_wave = trimed_ref_wave[0:trimed_wave_ch_2_3.shape[0],:]
+    compare_result2_3 = np.array_equal(trimed_ref_wave, trimed_wave_ch_2_3)
+
+    smart_amp_delay = ((delay2 - delay1) / fs * 1000)
+    print("Delay of smart amplifier is %0.3fms" % smart_amp_delay)
+    if compare_result0_1 and compare_result2_3 and smart_amp_delay < 5.:
+        print('Recorded wave is binary same as reference wave')
+        print("Wave comparison result: PASSED")
+    else:
+        print('Recorded wave is not binary same as reference wave')
+        print("Wave comparison result: FAILED")
+        sys.exit(1001)
+
+def parse_cmdline():
+    parser = argparse.ArgumentParser(add_help=True, formatter_class=argparse.RawTextHelpFormatter,
+        description='A Tool to Generate and Manipulate Wave Files.')
+    parser.add_argument('-v', '--version', action='version', version='%(prog)s 1.0')
+    # wave parameters
+    parser.add_argument('-g', '--generate', type=str, choices=['sine','cosine'], help='generate specified types of wave')
+    parser.add_argument('-A', '--amp', type=float, nargs='+', default=[1.0], help='amplitude of generated wave')
+    parser.add_argument('-F', '--freq', type=float, nargs='+', default=[997.0], help='frequency of generated wave')
+    parser.add_argument('-P', '--phase', type=float, nargs='+', default=[0.0], help='phase of generated wave')
+    parser.add_argument('-D', '--duration', type=float, nargs='+', default=[10.], help='duration of generated wave')
+    parser.add_argument('-S', '--sample_rate', type=int, default=48000, help='sample rate of generated wave')
+    parser.add_argument('-C', '--channel', type=int, default=2, help='channels of generated wave')
+    parser.add_argument('-B', '--bits', type=str, choices=['S8', 'S16', 'S32', 'F32'], default='S16',
```

Comment:
Could we re-use the numpy names here? I mean `int8`, `int16`,...

Reply:
If we want to reuse these value, we need a meta class, seems there is not such a class.

Reply:
I meant this:
```python
parser.add_argument('-B', '--bits', type=str, choices=['int8', 'int16', 'in32', ...], default='S16',
```
No real code change, only a string change for consistency with numpy.

Reply:
The `parse_val()` do make code shorter in this case, but since all of our case using this style, we may consider to do overall clean up, rather than only do it in this case.  A `parse_val` in this case may cause this function duplicated in other cases.   I think we may rename "func_pipeline_parse_value" to be shorter.